File: backend/render/edl.py
"""The Edit Decision List — the compact, typed contract between the director
agent and the renderer. One Shot per directorial decision; the renderer turns
each into a virtual-camera move over the matted composite, then stitches them
with the given transitions and per-scene grade.

Times are in *source* seconds. Shots are expected in source-time order and
contiguous (coverage-cut of a single take); the assembler keeps the dialogue
audio continuous across cuts.
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field, asdict
from typing import Any, Dict, List, Optional

SHOTS = ("WIDE", "MS", "MCU", "CU", "OTS")
SUBJECTS = ("A", "B", "both", "none")          # A = performer on LEFT, B = RIGHT
LOOKS = ("Neutral", "Noir", "Sci-Fi", "Golden", "Thriller")
TRANSITIONS = ("cut", "fade", "dissolve", "wipeleft", "wiperight", "circleopen")

# Perception agent vocab -> renderer vocab (so a director-produced read drops in)
_LOOK_ALIASES = {"neutral": "Neutral", "noir": "Noir", "sci-fi": "Sci-Fi",
                 "scifi": "Sci-Fi", "golden": "Golden", "thriller": "Thriller"}
_SHOT_ALIASES = {"wide": "WIDE", "ms": "MS", "medium": "MS", "mcu": "MCU",
                 "cu": "CU", "closeup": "CU", "close-up": "CU", "ots": "OTS"}


def _clamp_enum(value: Optional[str], allowed, default: str, aliases=None) -> str:
    if value is None:
        return default
    v = str(value).strip()
    if aliases and v.lower() in aliases:
        return aliases[v.lower()]
    return v if v in allowed else default


@dataclass
class Shot:
    id: str
    start: float                       # source seconds, inclusive
    end: float                         # source seconds, exclusive
    shot: str = "WIDE"                 # framing (SHOTS)
    subject: str = "both"             # who to frame on (SUBJECTS)
    look: str = "Neutral"             # per-scene grade (LOOKS)
    transition_in: str = "cut"        # how this shot begins vs the previous (TRANSITIONS)
    transition_dur: float = 0.4        # seconds; ignored for "cut"
    move: Optional[str] = None         # optional camera override: push|static|pan_left|pan_right
    background: Optional[Dict[str, Any]] = None  # {"file":..} | {"prompt":..} | {"color":[r,g,b]} | None
    note: str = ""                     # director_note, carried for overlays/debug

    def __post_init__(self):
        self.start = float(self.start)
        self.end = float(self.end)
        if self.end <= self.start:
            raise ValueError(f"shot {self.id!r}: end ({self.end}) must be > start ({self.start})")
        self.shot = _clamp_enum(self.shot, SHOTS, "WIDE", _SHOT_ALIASES)
        self.subject = _clamp_enum(self.subject, SUBJECTS, "both")
        self.look = _clamp_enum(self.look, LOOKS, "Neutral", _LOOK_ALIASES)
        self.transition_in = _clamp_enum(self.transition_in, TRANSITIONS, "cut")
        self.transition_dur = max(0.0, float(self.transition_dur))

    @property
    def duration(self) -> float:
        return self.end - self.start


@dataclass
class EDL:
    clip: str = ""                     # source video path (CLI --clip overrides)
    width: int = 1280
    height: int = 720
    fps: Optional[float] = None        # output fps; None -> source fps
    shots: List[Shot] = field(default_factory=list)
    meta: Dict[str, Any] = field(default_factory=dict)

    @property
    def size(self):
        return (int(self.width), int(self.height))

    def validate(self) -> "EDL":
        if not self.shots:
            raise ValueError("EDL has no shots")
        if self.width <= 0 or self.height <= 0:
            raise ValueError(f"bad output size {self.width}x{self.height}")
        # first shot can't dissolve from nothing
        self.shots[0].transition_in = "cut"
        return self

    def to_json(self, indent=2) -> str:
        d = asdict(self)
        return json.dumps(d, indent=indent)


def from_dict(d: Dict[str, Any]) -> EDL:
    shots = [Shot(**s) if not isinstance(s, Shot) else s for s in d.get("shots", [])]
    return EDL(
        clip=d.get("clip", ""),
        width=int(d.get("width", 1280)),
        height=int(d.get("height", 720)),
        fps=d.get("fps"),
        shots=shots,
        meta=d.get("meta", {}),
    ).validate()


def load(path: str) -> EDL:
    with open(path, "r") as f:
        return from_dict(json.load(f))

File: web/lib/edl.ts
// Cut! — the Edit Decision List, shared "final clip" contract.
//
// This is the TypeScript mirror of the renderer's typed contract in
// backend/render/edl.py — SAME field names, SAME vocabulary, SAME defaults — so an
// EDL produced here in the browser drops straight into `edl.from_dict(...)` on the
// render service (POST /render {edl, clip}). backend/render/edl.py is the source of
// truth; keep the enums below in lock-step with it.
//
// One Shot per directorial decision. Times are in *source* seconds (relative to the
// recorded take), in order and contiguous — a coverage-cut of a single take.

export const SHOTS = ["WIDE", "MS", "MCU", "CU", "OTS"] as const;
export const SUBJECTS = ["A", "B", "both", "none"] as const; // A = performer on LEFT, B = RIGHT
export const LOOKS = ["Neutral", "Noir", "Sci-Fi", "Golden", "Thriller"] as const;
export const TRANSITIONS = ["cut", "fade", "dissolve", "wipeleft", "wiperight", "circleopen"] as const;

export type ShotKind = (typeof SHOTS)[number];
export type Subject = (typeof SUBJECTS)[number];
export type Look = (typeof LOOKS)[number];
export type Transition = (typeof TRANSITIONS)[number];

// Mirrors render/edl.py:Shot (field names must match exactly).
export type Shot = {
  id: string;
  start: number; // source seconds, inclusive
  end: number; // source seconds, exclusive (must be > start)
  shot: ShotKind;
  subject: Subject;
  look: Look;
  transition_in: Transition;
  transition_dur: number; // seconds; ignored for "cut"
  move?: string; // push|static|pan_left|pan_right
  background?: { file?: string; prompt?: string; color?: [number, number, number] } | null;
  note: string;
};

// Mirrors render/edl.py:EDL.
export type Edl = {
  clip: string; // source path/URL; may be "" — the render POST carries `clip` separately
  width: number;
  height: number;
  fps: number | null;
  shots: Shot[];
  meta: Record<string, unknown>;
};

// A caption cue as the Audition engine records them: a co-star speech span in
// *recorded-media* seconds. (Actor lines produce no cue — they are the gaps.)
export type Cue = { t: number; end: number; text: string; who: string };

export type TakeToEdlOpts = {
  durationSec: number; // total recorded-media length of the take
  look?: Look; // per-scene grade (default Neutral)
  width?: number;
  height?: number;
  fps?: number;
  meta?: Record<string, unknown>;
};

// A loose tone → look mapping so a scene's mood picks a sensible grade. Deterministic;
// unknown tones fall back to Neutral. Callers may override with opts.look.
export function lookForTone(tone: string | undefined): Look {
  const t = (tone || "").toLowerCase();
  if (/noir|shadow|smok|bruised|jaded/.test(t)) return "Noir";
  if (/thrill|combust|tense|high-stakes|contempt|danger/.test(t)) return "Thriller";
  if (/golden|warm|tender|aching|nostalg/.test(t)) return "Golden";
  if (/sci-?fi|cold|clinical|glacial|icy/.test(t)) return "Sci-Fi";
  return "Neutral";
}

const EPS = 0.08; // ignore sub-frame slivers (~2 frames @ 30fps)

// Turn a recorded take's cue track into a valid, contiguous EDL the renderer accepts.
//
// Cues mark the co-star's speech spans; the gaps between them (and the lead-in / tail)
// are the actor's turns. We tile the whole [0, durationSec] timeline into alternating
// segments — actor → subject "A" (framed MS), co-star → subject "B" (framed MCU) — so
// coverage is complete, source-ordered, and every shot has end > start.
export function takeToEdl(cues: Cue[], opts: TakeToEdlOpts): Edl {
  const look: Look = opts.look ?? "Neutral";
  const duration = Math.max(0, opts.durationSec);
  const width = opts.width ?? 1280;
  const height = opts.height ?? 720;
  const fps = opts.fps ?? 30;

  type Seg = { start: number; end: number; subject: Subject; note: string };
  const segs: Seg[] = [];
  const sorted = [...cues].sort((a, b) => a.t - b.t);
  let cursor = 0;

  const pushActor = (start: number, end: number) => {
    if (end - start > EPS) segs.push({ start, end, subject: "A", note: "" });
  };

  for (const cue of sorted) {
    const cs = Math.max(cursor, Math.min(cue.t, duration));
    // a well-formed cue has end > t; interrupted cues (end 0) get a short default beat
    const rawEnd = cue.end > cue.t ? cue.end : cue.t + 2;
    const ce = Math.min(Math.max(rawEnd, cs), duration);
    pushActor(cursor, cs); // actor turn before this co-star line
    if (ce - cs > EPS) segs.push({ start: cs, end: ce, subject: "B", note: cue.text });
    cursor = Math.max(cursor, ce);
  }
  pushActor(cursor, duration); // trailing actor beat / silence

  // Degenerate take (no usable spans): one WIDE covering whatever we have.
  if (segs.length === 0 && duration > EPS) {
    segs.push({ start: 0, end: duration, subject: "both", note: "" });
  }

  const shots: Shot[] = segs.map((s, i) => ({
    id: `s${i}`,
    start: round(s.start),
    end: round(s.end),
    shot: s.subject === "B" ? "MCU" : s.subject === "both" ? "WIDE" : "MS",
    subject: s.subject,
    look,
    transition_in: "cut", // first shot must be a cut anyway; keep dialogue continuous
    transition_dur: 0.4,
    note: s.note,
  }));

  return {
    clip: "",
    width,
    height,
    fps,
    shots,
    meta: { source: "audition", ...(opts.meta ?? {}) },
  };
}

function round(x: number): number {
  return Math.round(x * 1000) / 1000;
}
